solved by **glpk**

In [64]:
!pip install -q pyomo 

In [65]:
from pyomo.environ import *

In [66]:
model1=ConcreteModel()


In [67]:
import numpy as np

In [68]:
N=3
M=3
lb=np.array([2,1,1])
ub=np.array([np.inf,np.inf,np.inf])
obj_coef_c= np.array([2,-2,2])
constr_coef_A= np.array([[2,-1,2],[1,-4,2],[-2,2,1]])
constr_rhs_b=np.array([6,-15,-8])
row_indices=np.arange(M)
print(row_indices)
col_indices=np.arange(N)
print(col_indices)

[0 1 2]
[0 1 2]


In [69]:
model1.x=Var(col_indices)
model1.constraints=ConstraintList()

In [70]:
for i in col_indices:
  model1.x[i].setlb(lb[i])
  model1.x[i].setub(ub[i])

In [71]:
for i in row_indices:
      model1.constraints.add(sum(constr_coef_A[i][j]*model1.x[j] for j in col_indices) <= constr_rhs_b[i])

In [72]:
model1.objective = Objective(expr =summation(obj_coef_c,model1.x), sense=minimize)
model1.pprint() 

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.

In [73]:
!apt-get install -y -qq glpk-utils

In [74]:
opt = SolverFactory('glpk', executable='/usr/bin/glpsol')
#.solve(model1).write()
result = opt.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 10
  Sense: minimize
Solver: 
- Status: ok
  Termination condition: other
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.01510000228881836

Solver status: ok
Solver termination condition: other


solved by cbc

In [75]:
!apt-get install -y -qq coinor-cbc

In [76]:
opt_cbc = SolverFactory('cbc')

In [77]:
result = opt_cbc.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.020425796508789062

Solver status: warning
Solver termination condition: infeasible


glpk solver  is fast by cbc solver , and glpk Solver termination condition not meaningfull define but cbc solver is define what is solver termination condition all says that [fesible yeah not fesible] that reson we are to choose cbc solver


Change the optimization problem to a maximization problem and re-solve.

In [78]:
model1.objective.set_sense(maximize)

In [79]:
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.

In [80]:
result = opt_cbc.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: -inf
  Upper bound: None
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: maximize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.033754587173461914

Solver status: warning
Solver termination condition: infeasible


not fesible solution 

Add upper bounds on x2, x3 as x2 ≤ 5, x3 ≤ 10 and re-solve

In [81]:
model1.x[1].setub(5)
model1.x[2].setub(10)
model1.pprint()
model1.objective.set_sense(minimize)

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :     5 : False :  True :  Reals
          2 :     1 :  None :    10 : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=3, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :   6.

In [82]:
result = opt_cbc.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: inf
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.03880190849304199

Solver status: warning
Solver termination condition: infeasible



they are infisible solution

c) Add a new constraint −x2 + x3 ≤ 16 and re-solve


In [83]:
model1.x[1].setub(np.inf)
model1.x[2].setub(np.inf)


In [84]:
model1.constraints.add(expr = -model1.x[1]+model1.x[2] <=16)

In [85]:
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    4 : {1, 2, 3, 4}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints : Size=4, Index=constraints_index, Active=True
        Key : Lower : Body                    : Upper : Active
          1 :  -Inf :  2*x[0] - x[1] + 2*x[2] :  

In [86]:
result = opt_cbc.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

  - termination condition: infeasible
  - message from solver: <undefined>



Problem: 
- Name: unknown
  Lower bound: None
  Upper bound: 11.0
  Number of objectives: 1
  Number of constraints: 5
  Number of variables: 4
  Number of nonzeros: 0
  Sense: minimize
Solver: 
- Status: warning
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: infeasible
  Termination message: Model was proven to be infeasible.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
    Black box: 
      Number of iterations: 0
  Error rc: 0
  Time: 0.02599334716796875

Solver status: warning
Solver termination condition: infeasible


infisible solution

(d) Deactivate the objective of problem (OP) and construct a new objective of the form
min x1 + 2x2 + 3x3 and deactivate the constraint 2x1 − x2 + 2x3 ≤ 6 and add a different
constraint x2 + x3 ≥ 25

In [87]:
new_constr_coef =[0,1,1]
new_constr_rhs=25
model1.constraints.add(sum(new_constr_coef[j]*model1.x[j] for j in col_indices) >= new_constr_rhs)



In [88]:
model1.constraints[1].deactivate()
model1.constraints[4].deactivate()

model1.objective.deactivate()

model1.new_objective = Objective(expr = model1.x[0 ]+ 2*model1.x[1]+3*model1.x[2])


In [89]:
model1.objective.set_sense(minimize)

In [90]:
model1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {0, 1, 2}

1 Var Declarations
    x : Size=3, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     2 :  None :  None : False :  True :  Reals
          1 :     1 :  None :  None : False :  True :  Reals
          2 :     1 :  None :  None : False :  True :  Reals

2 Objective Declarations
    new_objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : x[0] + 2*x[1] + 3*x[2]
    objective : Size=1, Index=None, Active=False
        Key  : Active : Sense    : Expression
        None :  False : minimize : 2*x[0] - 2*x[1] + 2*x[2]

1 Constraint Declarations
    constraints

In [91]:
!apt-get install -y -qq coinor-cbc


In [92]:
opt_cbc = SolverFactory('cbc')

In [93]:
result = opt_cbc.solve(model1)
print(result)

print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)
print('Solver time:',result.solver.time)



Problem: 
- Name: unknown
  Lower bound: 79.5
  Upper bound: 79.5
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 4
  Number of nonzeros: 3
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 2
  Error rc: 0
  Time: 0.022576332092285156
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

Solver status: ok
Solver termination condition: optimal
Solver time: 0.022576332092285156


optimal solution

In [94]:
#display solution
print('\nObjective = ', model1.new_objective())

print('\nDecision Variables')
for i in col_indices:
  print('x'+str(i+1)+'= ', model1.x[i].value)
 
 
print('\nConstraints')
model1.constraints.display()


Objective =  79.5

Decision Variables
x1=  28.5
x2=  24.0
x3=  1.0

Constraints
constraints : Size=5
    Key : Lower : Body  : Upper
      2 :  None : -65.5 : -15.0
      3 :  None :  -8.0 :  -8.0
      5 :  25.0 :  25.0 :  None


the minimum value are  =  79.5

Decision Variables
x1=  28.5
x2=  24.0
x3=  1.0
consrint 1 are inactive, and constarint 2 are in active, costraint 3 are active